### DISCUSSION TITLE
Creating a simple web app

### DISCUSSION THREAD

**Discussion by fdinhSG** [pyvista/pyvista#4297] [Q&A] [answered]
Question related to https://github.com/zauberzeug/nicegui/discussions/745

Reference code:
```python
import pyvista, os
from nicegui import app, ui

pl = pyvista.Plotter(window_size=(800, 600))
mesh_actor = None
def create_mesh(solid):
    global pl, mesh_actor
    if mesh_actor is not None:
        pl.remove_actor(mesh_actor)
    match solid:
        case 'sphere':
            mesh_actor = pl.add_mesh(pyvista.Sphere())
        case 'cube':
            mesh_actor = pl.add_mesh(pyvista.Cube())
    tmp = os.path.join(os.getcwd(),'tmp')
    pl.export_html(os.path.join(tmp,'mesh.html'), backend='panel')
    app.add_static_files('/meshes', tmp)
    iframe.content = '<iframe src="/meshes/mesh.html" width="800" height="600"></iframe>'  # mesh.html is different, but iframe.content doesn't get updated!

iframe = ui.html('<iframe width="800" height="600"></iframe>')
with ui.row():
    ui.button('Sphere', on_click=lambda: create_mesh('sphere'))
    ui.button('Cube', on_click=lambda: create_mesh('cube'))

ui.run()
```

I want to create different meshes and visualize them on an iframe. I can change the html element content in order to view the meshes, however since the new mesh has the same name as the one before (mesh.html), the page won't reload and see the changes. Is there a way to force a page reload every time I have a new mesh.html?

Another question: since I need to delete the previous mesh actor on the plot I have to return this value from the `create_mesh` function. Since I wasn't able to return this value I had to resort to global variables which I don't particularly like. Is there a way to return a variable from the `create_mesh` function and assign it to a variable somehow?

Thanks!

**Comment by banesullivan** [comment, marked as answer]
You should try out [Trame](https://kitware.github.io/trame/index.html) with PyVista -- your current example with static exports of the plots will not scale or update in a reactive way. I realize this is about to sound terse, but it's simply the wrong approach... you need a reactive framework that will update the content on the page. There are many options (`dash` for one), but we are full-heartedly behind using Trame with PyVista.

We have some examples with Trame here: https://github.com/pyvista/pyvista/tree/main/examples_trame

Here is your example as a trame app:

```py
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import vuetify

import pyvista as pv
from pyvista.trame.ui import plotter_ui

# -----------------------------------------------------------------------------
# Trame initialization
# -----------------------------------------------------------------------------

pv.OFF_SCREEN = True

server = get_server()
state, ctrl = server.state, server.controller

state.trame__title = "My App"
ctrl.on_server_ready.add(ctrl.view_update)


# -----------------------------------------------------------------------------
# Plotting
# -----------------------------------------------------------------------------

pl = pv.Plotter(window_size=(800, 600))
mesh_actor = None


def create_mesh(solid):
    global pl, mesh_actor
    if mesh_actor is not None:
        pl.remove_actor(mesh_actor)
    if solid == 'sphere':
        mesh_actor = pl.add_mesh(pv.Sphere())
    elif solid == 'cube':
        mesh_actor = pl.add_mesh(pv.Cube())
    pl.reset_camera()


# -----------------------------------------------------------------------------
# GUI
# -----------------------------------------------------------------------------


with SinglePageLayout(server) as layout:
    layout.title.set_text(state.trame__title)

    with layout.toolbar:
        vuetify.VSpacer()

        vuetify.VBtn("Sphere", click=lambda: create_mesh("sphere"))

        vuetify.VBtn("Cube", click=lambda: create_mesh("cube"))

    with layout.content:
        with vuetify.VContainer(
            fluid=True,
            classes="pa-0 fill-height",
        ):
            # Use PyVista UI template for Plotters
            view = plotter_ui(pl)
            ctrl.view_update = view.update

server.start()
```

**Reply by banesullivan** [reply]
I want to add that I know nothing about `nicegui` and it very well could be a compatible framework. There has been a ton of work into getting VTK/PyVista to work optimally on the web in Trame and I suspect Trame is the best route forward if you want reactivity. Hope this helps!

**Reply by banesullivan** [reply]
Please also note that you could have a trame app like this for the PyVista plotter embedded as an iframe in your nicegui project if you are already invested in that tooling

**Reply by fdinhSG** [reply]
Hi @banesullivan, thank you for your reply! I wasn't aware of trame integration - i took the export_html approach from [the flask example](https://docs.pyvista.org/version/stable/extras/flask.html) and I thought that was the "standard" approach. I'll gladly explore the trame option!

On a related note, it looks like I'd be able to select points on the mesh using trame and have more interactivity than with a standard iframe element, is that correct? If so I'll probably switch right away. Thanks again!

**Reply by banesullivan** [reply]
> i took the export_html approach from [the flask example](https://docs.pyvista.org/version/stable/extras/flask.html) and I thought that was the "standard" approach.

Ah, thanks for bringing this to my attention -- I will work to clarify this and perhaps remove the flask example as that was a rough proof of concept some time ago

**Reply by banesullivan** [reply]
> On a related note, it looks like I'd be able to select points on the mesh using trame and have more interactivity than with a standard iframe element, is that correct? If so I'll probably switch right away. Thanks again!

Correct, some, but not all of the picking will work

**Reply by fdinhSG** [reply]
> Correct, some, but not all of the picking will work

This is great. I'll try it asap and play with it, thanks

**Reply by banesullivan** [reply]
FYI, the picking is undergoing an overhaul in https://github.com/pyvista/pyvista/pull/4254

**Reply by akaszynski** [reply]
We should consider removing the flask example.

**Comment by banesullivan** [comment]
> Another question: since I need to delete the previous mesh actor on the plot I have to return this value from the create_mesh

You can actually use the `name` argument in `add_mesh()` to handle all of this for you. Actors must have unique names. So if you add a mesh with a name that is already present it will be replaced. Looks like:

```py
def create_mesh(solid):
    if solid == 'sphere':
        mesh = pv.Sphere()
    elif solid == 'cube':
        mesh = pv.Cube()
    pl.add_mesh(mesh, name='my mesh')
    pl.reset_camera()
```
